crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs

```diff
@@ -1,22 +1,158 @@
+    value: Expression,
+    index: Expression,
+}
+
 #[derive(Default)]
 pub struct ArrayHeap {
     // maps memory address to InternalVar
     memory_map: BTreeMap<u32, InternalVar>,
+    trace: Vec<MemItem>,
+    staged: BTreeMap<u32, (Expression, Expression)>,
+}
+
+impl ArrayHeap {
+    pub fn commit_staged(&mut self) {
+        for (idx, (value, op)) in &self.staged {
+            let item = MemItem {
+                op: op.clone(),
+                value: value.clone(),
+                index: Expression::from_field(FieldElement::from(*idx as i128)),
+            };
+            self.trace.push(item);
+        }
+        self.staged.clear();
+    }
+
+    pub fn push(&mut self, index: Expression, value: Expression, op: Expression) {
+        let item = MemItem { op, value, index };
+        self.trace.push(item);
+    }
+
+    pub fn stage(&mut self, index: u32, value: Expression, op: Expression) {
+        self.staged.insert(index, (value, op));
+    }
+
+    pub fn acir_gen(&self, evaluator: &mut Evaluator) {
+        let len = self.trace.len();
+        if len == 0 {
+            return;
+        }
+        let len_bits = AcirMem::bits(len);
+        // permutations
+        let mut in_counter = Vec::new();
+        let mut in_index = Vec::new();
+        let mut in_value = Vec::new();
+        let mut in_op = Vec::new();
+        let mut out_counter = Vec::new();
+        let mut out_index = Vec::new();
+        let mut out_value = Vec::new();
+        let mut out_op = Vec::new();
+        let mut tuple_expressions = Vec::new();
+        for (counter, item) in self.trace.iter().enumerate() {
+            let counter_expr = Expression::from_field(FieldElement::from(counter as i128));
+            in_counter.push(counter_expr.clone());
+            in_index.push(item.index.clone());
+            in_value.push(item.value.clone());
+            in_op.push(item.op.clone());
+            out_counter.push(expression_from_witness(evaluator.add_witness_to_cs()));
+            out_index.push(expression_from_witness(evaluator.add_witness_to_cs()));
+            out_value.push(expression_from_witness(evaluator.add_witness_to_cs()));
+            out_op.push(expression_from_witness(evaluator.add_witness_to_cs()));
+            tuple_expressions.push(vec![item.index.clone(), counter_expr.clone()]);
+        }
+        let bit_counter =
+            operations::sort::evaluate_permutation(&in_counter, &out_counter, evaluator);
+        operations::sort::evaluate_permutation_with_witness(
+            &in_index,
+            &out_index,
+            &bit_counter,
+            evaluator,
+        );
+        operations::sort::evaluate_permutation_with_witness(
+            &in_value,
+            &out_value,
+            &bit_counter,
+            evaluator,
+        );
+        operations::sort::evaluate_permutation_with_witness(
+            &in_op,
+            &out_op,
+            &bit_counter,
+            evaluator,
+        );
```

Comment:
This whole chunk is very repetitive, we should factor it out

Reply:
Done